Support for reading and writing sequencer states

bpeabody opened the pull request:
Addreses task: https://github.com/twosigma/git-meta/projects/2#card-8117068

novalis commented on node/lib/util/sequencer_state.js:
```diff
@@ -0,0 +1,156 @@
+/**
+ * @class CommitAndRef
+ *
+ * This class describes a commit and optionally the ref it came from.
```
Isn't that just a NodeGit.AnnotatedCommit?  Maybe there's no way to get the ref out of that, tho.

bpeabody replied:
Yes, that's what I'd use ideally, but there's no way to make an `AnnotatedCommit` without a repo (which makes it harder to test with them, and as you point out, no way to get the ref back.

novalis commented on node/lib/util/sequencer_state_util.js:
```diff
@@ -0,0 +1,237 @@
+    if (null !== content) {
+        const lines = content.split("\n");
+
+        // always an extra entry after the split
```
Do you mean that the file consists of x\ny\n, and so split gives ['x', 'y', '']?  If so, strip the trailing newline before split to make clearer what's going on.

novalis commented on node/lib/util/sequencer_state_util.js:
```diff
@@ -0,0 +1,237 @@
+});
+
+/**
+ * Return the array of commit stored in the specified `gitDir`, or null if
```
Can you describe a bit better what this file is?  All I know now is that it's a list of commits -- but what for?

novalis replied:
Actually, maybe describe the directory schema at the top of the file, because I have no idea why we are reading these files or whatever.

bpeabody replied:
good idea

novalis commented on node/lib/util/sequencer_state_util.js:
```diff
@@ -0,0 +1,237 @@
+ * @param {Number} numCommits
+ * @return {Number}
+ */
+exports.readCurrentCommit = co.wrap(function *(gitDir, numCommits) {
```
I don't understand what numCommits means or what values are valid.

novalis commented on node/lib/util/sequencer_state_util.js:
```diff
@@ -0,0 +1,237 @@
+/**
+ * Return the sequencer state if it exists in the specified `gitDir`, or null
+ * if it is missing or malformed.
+ * TODO: emit diagnostic when malformed?
```
raise exception if malformed -- let the caller say, "hey, your sequencer state is malformed.  This shouldn't happen.  You can fix it by deleting .git/meta-sequencer and then using the reflog to get back to where you started."

(ok to leave like this for this pr)

bpeabody replied:
Yes, I want to think about this some more.

novalis commented on node/lib/util/sequencer_state_util.js:
```diff
@@ -0,0 +1,237 @@
+    if (null === currentCommit) {
+        return null;                                                  // RETURN
+    }
+    return new SequencerState(type, original, target, commits, currentCommit);
```
Would a fluent builder pattern be better here, or is that not a thing in JS?

bpeabody replied:
To deal reduce the number of args? I think what you'd be more likely to see in JS would be a dictionary, especially if you have a optional arguments. I didn't use one because all the args here are required, but it would look like:
```JavaScript
SequencerState({
    type: type
    original: original,
    target: target,
    commits: commits,
    currentCommit: currentCommit,
});
```

novalis replied:
Not just reduce the number, but also avoid having to look up ordering.  I like that version better.

bpeabody replied:
I'm fine with that. I could see this constructor getting more args in the future, and I don't think I'd go to 6.

bpeabody commented:
OK, let's see if this version is more comprehensible.

novalis commented on node/lib/util/sequencer_state_util.js:
```diff
@@ -0,0 +1,254 @@
+ * .git/meta_sequencer/TYPE            -- Single line containing the type, e.g.
+ *                                        "REBASE".
+ *                    /ORIGINAL_HEAD   -- SHA and optional ref name of what was
+ *                                     -- on head.  If first line is the SHA.
```
sentence fragment "If the first line is the SHA" is not a syntactic constituent.

bpeabody commented:
OK, I think I addressed all the concerns, but I also slipped in a couple of methods that I will need but had forgotten: `equal` and `toString` if you want to take a quick look at them, @novalis .

bpeabody commented on node/lib/util/sequencer_state.js:
```diff
@@ -0,0 +1,210 @@
+     * @param {CommitAndRef} rhs
+     * @return {Bool}
+     */
+    equal(rhs) {
```
I think the types they're referring to in that doc (e.g., `Hash`) are a library that's not part of standard JS, which, unfortunately doesn't seem to have standard operators that I've seen for equality testing and what not. In any case, I made an `equal`, not an `equals`.

Either way, I've been using `equal` in other classes. If this code is OK otherwise, I'd be happy to go back and revisit across the board later.

novalis commented:
Oy, Javascript.
